### DISCUSSION TITLE
transitions of different source_ranges' rates and movies' fps

### DISCUSSION THREAD

**Discussion by ggarra13** [AcademySoftwareFoundation/OpenTimelineIO#1320] [Q&A] [answered]
My player (mrViewer) currently works with frames, not time.  Movies and sequences have fps control for playback ( and a play_fps to play at different speed ).

My question is how can I deal with different rates in source_ranges when they do not match the fps of the movie files.

Currently, when transitioning from a movie with a certain fps ( say 30 ) and the same rate in the source range and a transition also of rate 30 and another movie of fps ( say 23.976 ) and its source range of rate 30, too, I seem like I need to adjust the start and end frames to compensate for the source range rate (which overrides the movies' fps).  That works fine.  

However, when the transition and playback is involved, the movie fps of the files seems like it needs to be respected (am I right?) and thus I have to play back the two clips at different fps.  
The problem is that when the transition ends, the frame where it ends in the movie file of fps of 23.976 no longer matches the source range of rate 30 and thus I get an offset in the timeline.

If I force the clip of fps 23.976 to play at a rate 30 then the timeline and transition matches, but of course my clip plays too fast.

Hope this makes sense.

**Comment by jminor** [comment, marked as answer]
The rate of a Clip's source_range and the rate of the Media Reference don't need to match. If you want your composition to have a consistent rate, then you can ensure that all of your Clips' source_ranges are all using the desired rate. When the media doesn't match that, then it is up to the playback tool or renderer to determine what to do. Depending on the context and the desired outcome, it might make sense to resample the media (dropping or repeating frames as needed) or do some blending operation. Different NLEs handle this in different ways, so OTIO doesn't force one way of resolving it.

In systems that are completely frame-based, the common strategy is to conform all input media to a consistent rate so that everything lines up on frames before composition is done (e.g. the transition case you described). In systems that are time-based then the media can be sampled at the output rate - similar to spatial sampling when stretching a bitmap to fit a different resolution.

This temporal re-sampling may already be happening in many programs due to monitor refresh rates. For example, if you play 24fps media on a 90Hz monitor, then some frames are shown longer than others (since 90/24 = 3.75), but on a 72Hz monitor they are consistent (since 72/24 = 3)

---
Joshua Minor (he-him)
Pixar Studio Tools Story/Editorial/VR Previs Tech Lead
***@***.***

**Comment by jminor** [comment]
@ggarra13 did that answer help resolve your question about varying rates?

**Reply by ggarra13** [reply]
Yes.  Sadly mrViewer's architecture is currently not set up for doing that.  I am rewriting mrViewer now from scratch (to clean all the spaguetti code and all the LICENSE issues) with Darby's excellent tlRender library.

**Reply by jminor** [reply]
Oh wow! Looking forward to seeing where that leads :)

**Reply by ggarra13** [reply]
mrViewer 2 (mrv2) is now in alpha.  You can find the repository and instructions to compile at:  https:/github.com/ggarra13/mrv2 and binaries (Windows 8.1, macOS Mojave or later and Ubuntu 22.04 LTS --sorry for that) for the first alpha release there or on sourceforge.net/p/mrv2.
